### DISCUSSION TITLE
Is there a way to keep a method's definition after decoration?

### DISCUSSION THREAD

**Discussion by belegnar** [microsoft/pyright#8749] [Q&A] [answered]
Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwEYGMBQpKzjAnmYAOwHNEBnHNNFAGwTIoFUwATBGAUwC404%2B4w9MlVpC4AQSRkYUBChgAKZCgB04gEIBhAJQ9%2B5XIRRwWHAGYmIAfQBWEJArIcaZgDRwArq3bc4zNpzacAC0AHxwAHIQhL4qcSJ0DHAAkoTSCEYcACIQHFAAClAQMBAoEDQKMCqFxaXlurz89IbGphZWVigINDQdjs5unt6cXH7DHEFhkdGx8dSJFDl5NSVlFVUrdTQN%2Bs1GJuZwHV09fU4uVu7nZqOS0rLy7l4Bvv4%2Bk%2BFRMaNwAMQ4%2BF8wGIhGgHAA2lAOFgYABlAaaGhkfIIWQgDicKARBDogC6jX0fDiKgJcAAAlUIAA3PI0CAIFiktpHKzEDFnAbuIjpTK3KQyOQwdwQADuMSgozwYAhdwF8lxcAAPtMYnAALwqiYhcKpHkobK5ApFVblUbE0mkikqam0%2BmM-TMjpsmAclxctIwDL60ZfDjCsV5SWA8Gyh4wBXK33qzUfOAAIiWRtqazjZrmDsOTvZVn6brg3M9vIk-LDSs1-vFQelIZLgojmujvtjVUYhGA0XBusL%2BsTmzW7gThr75TjuL0hKJc2ZpjKshKUAUZg8RlGveNW1ja%2BTptJe1amasRExhG6uZuxfugqe41Gb0C2s144nCBFCGA8CXRjP15e2lJUJgDwoEIFkjzyE8aASMRknACpQ0FHZ%2BDJGdoHYaBdzIFoDgsFhrDsBxrh-Hxb3GWNfSfQlBAYNAgA)

After several attempts I've ended up with this playground, but it still doesn't work as expected: `Impl` is incompatible with `Abstract` in some way

```python
import abc
import typing as t

class Update:
    pass

class Abstract(abc.ABC):
    async def do_job(self, update: Update) -> None: ...

class InstanceDoerProtocol(t.Protocol):
    async def __call__(self, update: Update) -> None: ...

class DoerProtocol(t.Protocol):
    async def __call__(self_, self: Abstract, update: Update) -> None:  # type: ignore[reportSelfClsParameterName]
        ...
    @t.overload
    def __get__(self, instance: Abstract, owner: type[Abstract] | None = None) -> InstanceDoerProtocol: ...
    
    @t.overload
    def __get__(self, instance: None, owner: type[Abstract] | None = None) -> "DoerProtocol": ...

    def __get__(self, instance: Abstract | None, owner: type[Abstract] | None = None) -> t.Union[InstanceDoerProtocol, "DoerProtocol"]:
        ...

def decorator(func: DoerProtocol) -> DoerProtocol:
    async def __internal(self: Abstract, update: Update) -> None:
        await func(self, update)
    return __internal

class Impl(Abstract):
    @decorator
    async def do_job(self, update: Update) -> None:
        pass

```

**Comment by bschnurr** [comment, marked as answer]
see https://microsoft.github.io/pyright/#/typed-libraries?id=annotating-decorators

**Reply by erictraut** [reply]
`Impl` is a subtype of `Abstract`, not a supertype. In a callable, parameter types are contravariant, so a compatible callable must have parameters that are supertypes. 

Here's a potential (although somewhat ugly) [solution](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwEYGMBQpKzjAnmYAOwHNEBnHNNFAGwTIoFUwATBGAUwC404%2B4w9MlVpC4AQSRkYUBChgAKZCgB04gEIBhAJQ9%2B5XIRRwWHAGYmIAfQBWEJArIcaZrnABEk6bPnuANHAArqzs3HDMbJzacAC0AHxwAHIQhGEq6SJ0DHAAkoTSCEYcACIQHFAAClAQMBAoEDQKMCpVNXUNurz89IbGphZWVigINDSDjs5mAcGRYRGh0fFJKWkZ1FkUpeWttfWNzTvtNJ36PUYm5nCDw6PjTi5WAfeuElIycjDTIZxu81GxCWSqTccAAxDh8GFgMRCNAOABtKAcLAwADKk00NDIFQQshAHE4UESCHxAF0uvo%2BOkVBS4AABZoQABu5RoEAQLFp-SuVmIBLukwCRAKRTcXne8gCEAA7qkoG48GAEeKfDBSXAAD7LVJwAC82o4iwSeRFKBKZUq1V2DTc1NptIZKmZrPZnP03MGfJgApcQvyMEKZrcQI4Utl5QVkPhKo%2B6q1Ib1BqNHi2lrae3ctrW7sunv5VgmvrgwoDote3g%2BmoNYblkaV0beqrjBsTIeTzUYhGAKXhJtLZtThz2AXcg6tR3cpL0lKpa25pnqslqUHhABU3GP0w1SQozIEjG5V8n17Szn1c1YiITCCNCy8Y5Kgt85s-kyHpzOENKEMB4HujHeXyzNotJIjAgRQIQOAqMM0gKKuASDFe5Q3scvBoI6ZhECM6xiDk4CNA%2BMAnPwdILtA7DQKeZC9BcFgsNYdgOM8YqNh8QGhL8r4AgaH6UoIDBoEAA).

Note that I've supplied explicit annotations for `self` in both `Abstract.do_job` and `Impl.do_job`. By annotating `self` as `Abstract` here rather than the default of `Self`, we avoid the contravariant issue noted above, but it makes `Impl` no longer usable as a base class. That's why I marked it as `@final`.

I also needed to add a `cast` call, which is a little distasteful but sometimes unavoidable.

**Reply by belegnar** [reply]
I can't realise how did I miss the docs, kindly pointed by @bschnurr, but it looks like working in the very [straightforward way](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwEYGMBQpKzjAnmYAOwHNEBnHNNFAGwTIoFUwATBGAUwC404%2B4w9MlVpC4AQSRkYUBChgAKZCgB04gEIBhAJQ9%2B5XIRRwWHAGYmIAfQBWEJArIcaZgDRwArq3bc4zNpzacAC0AHxwAHIQhL4qcVQAKnAAvDgqCfgcAGoIUAoARAn57kgQHoQsyTAqmgg0dEg0HADacSru1eIA7gjAMMhNzVExALoj2lSmFqYo0OzQCmblKFxwCUFha3r89IbGU3BWVkScUIR1js5mq5LSsvLuXgG%2B-j4b4cPcvPo7PX1wSyMlxcj28gW%2BfCgHBgHjOaRQ9EUCXcRxOHDOdQmcGodAYcAAkuAaApbjI5DBdBC4AABGZzGDQKm7IwmcyWWz2YFuTxgl6896RaJfH78QQMNBAA)
